pkg/cloud/affinity_groups_test.go

```diff
@@ -0,0 +1,119 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloud_test
+
+import (
+	"errors"
+
+	"github.com/apache/cloudstack-go/v2/cloudstack"
+	infrav1 "github.com/aws/cluster-api-provider-cloudstack/api/v1beta1"
+	"github.com/aws/cluster-api-provider-cloudstack/pkg/cloud"
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	capiv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+var _ = Describe("AffinityGroup Unit Tests", func() {
+	var ( // Declare shared vars.
+		mockCtrl    *gomock.Controller
+		mockClient  *cloudstack.CloudStackClient
+		ags         *cloudstack.MockAffinityGroupServiceIface
+		fakeAG      *cloud.AffinityGroup
+		cluster     *infrav1.CloudStackCluster
+		machine     *infrav1.CloudStackMachine
+		capiMachine *capiv1.Machine
+		client      cloud.Client
+	)
+
+	BeforeEach(func() {
+		// Setup new mock services.
+		mockCtrl = gomock.NewController(GinkgoT())
+		mockClient = cloudstack.NewMockClient(mockCtrl)
+		ags = mockClient.AffinityGroup.(*cloudstack.MockAffinityGroupServiceIface)
+		client = cloud.NewClientFromCSAPIClient(mockClient)
+		fakeAG = &cloud.AffinityGroup{
+			Name: "FakeAffinityGroup",
+			Type: cloud.AffinityGroupType}
+		cluster = &infrav1.CloudStackCluster{Spec: infrav1.CloudStackClusterSpec{
+			Zone: "Zone1", Network: "SharedGuestNet1"}}
+		machine = &infrav1.CloudStackMachine{Spec: infrav1.CloudStackMachineSpec{
+			Offering: "Medium Instance", Template: "Ubuntu20"}}
+		machine.ObjectMeta.SetName("rejoshed-affinity-group-test-vm")
+		capiMachine = &capiv1.Machine{}
+	})
+
+	AfterEach(func() {
+		mockCtrl.Finish()
+	})
+
+	It("fetches an affinity group", func() {
+		ags.EXPECT().GetAffinityGroupByName(fakeAG.Name).Return(&cloudstack.AffinityGroup{}, 1, nil)
+
+		Ω(client.GetOrCreateAffinityGroup(cluster, fakeAG)).Should(Succeed())
+	})
+	It("creates an affinity group", func() {
+		fakeAG.Id = "FakeID"
+		cluster.Spec.Account = "FakeAccount"
+		cluster.Status.DomainID = "FakeDomainId"
+		ags.EXPECT().GetAffinityGroupByID(fakeAG.Id).Return(nil, -1, errors.New("FakeError"))
+		ags.EXPECT().NewCreateAffinityGroupParams(fakeAG.Name, fakeAG.Type).
+			Return(&cloudstack.CreateAffinityGroupParams{})
+		ags.EXPECT().CreateAffinityGroup(ParamMatch(And(AccountEquals("FakeAccount"), DomainIdEquals("FakeDomainId")))).
+			Return(&cloudstack.CreateAffinityGroupResponse{}, nil)
+
+		Ω(client.GetOrCreateAffinityGroup(cluster, fakeAG)).Should(Succeed())
+	})
+
+	Context("AffinityGroup Integ Tests", func() {
```

Comment:
I feel pretty cool about this test setup.

Essentially, the integ suite is just skipped if there is no cloud config to work with. This way I can run integ tests locally, and eventually against our ACS setup too, without breaking things in the meantime.

Reply:
How do we avoid unintentionally skipping these tests?

Reply:
Once we're setup to actually hit a CS instance, then we can simply remove the conditional.

I've in the past also failed if the passing test count wasn't sufficient, so that's a possibility.

Another is labeling these as semi-integ and then running them in a separate job.

Overall, there are a bunch of different ways to handle it once we even get that far.